src/main/java/com/google/devtools/build/lib/rules/proto/BazelProtoLibrary.java

```diff
@@ -33,15 +37,53 @@
   public ConfiguredTarget create(RuleContext ruleContext)
       throws ActionConflictException, RuleErrorException, InterruptedException {
     ProtoCommon.checkRuleHasValidMigrationTag(ruleContext);
-    ProtoInfo protoInfo =
-        ProtoCommon.createProtoInfo(
-            ruleContext,
-            ruleContext.getFragment(ProtoConfiguration.class).generatedProtosInVirtualImports());
+    ProtoCommon.checkSourceFilesAreInSamePackage(ruleContext);
+
+    ImmutableList<Artifact> originalDirectProtoSources =
+        ruleContext.getPrerequisiteArtifacts("srcs", Mode.TARGET).list();
+    ImmutableList<ProtoInfo> deps = ImmutableList.copyOf(
+            ruleContext.getPrerequisites("deps", TARGET, ProtoInfo.PROVIDER));
+    ImmutableList<ProtoInfo> exports = ImmutableList.copyOf(
+            ruleContext.getPrerequisites("exports", TARGET, ProtoInfo.PROVIDER));
+    String protoSourceRoot =
+        ruleContext
+            .getLabel()
+            .getPackageIdentifier()
+            .getRepository()
+            .getExecPath(
+                ruleContext
+                    .getAnalysisEnvironment()
+                    .getSkylarkSemantics()
+                    .experimentalSiblingRepositoryLayout())
+            .getPathString();
+    Artifact directDescriptorSet =
+        ruleContext.getGenfilesArtifact(
+            ruleContext.getLabel().getName() + "-descriptor-set.proto.bin");
+
+    boolean generatedProtosInVirtualImports =
+        ruleContext.getFragment(ProtoConfiguration.class).generatedProtosInVirtualImports();
+    ProtoCommon.Library library =
+        ProtoCommon.createLibraryWithVirtualSourceRootMaybe(
+            ruleContext, originalDirectProtoSources, generatedProtosInVirtualImports);
     if (ruleContext.hasErrors()) {
       return null;
     }
 
-    ProtoCompileActionBuilder.writeDescriptorSet(ruleContext, protoInfo, Services.ALLOW);
+    if (library == null) {
+      library = ProtoCommon.createLibraryWithoutVirtualSourceRoot(protoSourceRoot, originalDirectProtoSources);
+    }
+
+    ProtoInfo protoInfo =
+        ProtoCommon.createProtoInfo(
+            originalDirectProtoSources,
+            library.getSources(),
```

Comment:
nit: why not just pass in `Library`?

Reply:
Even better, then you can put `originalDirectProtoSources` and `directDescriptorSet` also into it, which makes it a nice self-contained "things about this particular `proto_library` rule" class.

Reply:
Do you mean add `originalDirectProtoSources` and `directDescriptorSet` to `library` ?